File: frontend/js/clientes.js
const token = localStorage.getItem("token");

if (!token) {
    mostrarMensagem("Faça login para acessar", "erro");
    window.location.href = "login.html";
}

function ir(pagina) {
    window.location.href = pagina;
}

function logout() {
    localStorage.removeItem("token");
    window.location.href = "login.html";
}

function editarCliente(id, nome, email, telefone) {
    document.getElementById("nome").value = nome;
    document.getElementById("email").value = email;
    document.getElementById("telefone").value = telefone;

    window.clienteEditando = id;

    document.getElementById("bntSalvar").innerText = "Atualizar";
}

function cancelarEdicao() {
    window.clienteEditando = null;

    document.getElementById("nome").value = "";
    document.getElementById("email").value = "";
    document.getElementById("telefone").value = "";
    document.getElementById("bntSalvar").innerText = "Salvar";
}

async function criarCliente() {
    const nome = document.getElementById("nome").value.trim();
    const email = document.getElementById("email").value.trim();
    const telefone = document.getElementById("telefone").value.trim();


    if (!nome || !email || !telefone) {
        mostrarMensagem("Preencha todos os campos", "erro");
        return;
    }

    let endpoint = "/clientes/";
    let method = "POST";

    if (window.clienteEditando) {
        endpoint += window.clienteEditando;
        method = "PUT";
    }

    try {

        await api(endpoint, {
            method: method,
            body: JSON.stringify({
                nome,
                email,
                telefone
            })
        });

        mostrarMensagem(window.clienteEditando ?
            "Cliente atualizado com sucesso!" : "Cliente criado com sucesso!", "sucesso");

        window.clienteEditando = null;

        document.getElementById("nome").value = "";
        document.getElementById("email").value = "";
        document.getElementById("telefone").value = "";
        document.getElementById("bntSalvar").innerText = "Salvar";

        buscar();

    } catch (error) {
        mostrarMensagem("Erro ao salvar cliente", "erro");
        console.error(error);
    }
}

async function buscar() {
    const nome = document.getElementById("busca").value.trim();
    const lista = document.getElementById("listar");

    lista.innerHTML = "";

    if (!nome) {
        lista.innerHTML = "<li>Digite um nome para buscar</li>";
        return;
    }

    const data = await api(`/clientes?nome=${nome}`);

    if (!data || data.length === 0) {
        lista.innerHTML = "<li>Nenhum cliente encontrado</li>";
        return;
    }

    data.forEach(c => {
        const tr = document.createElement("tr");
        tr.innerHTML = `<td>${c.nome}</td>
                        <td>${c.email}</td>
                        <td>${c.telefone}</td>
                <td>
                    <button onclick="editarCliente(
                        ${c.id},
                        '${c.nome}',
                        '${c.email}',
                        '${c.telefone}')">Editar</button>
                        
                    <button onclick="inativarCliente(
                        ${c.id})" class="btn-danger">Inativar</button>
                </td>`;

        lista.appendChild(tr);
    });
}

async function inativarCliente(id) {
    const confirmar = confirm("tem certeza que deseja inativar ? Você poderá reativar novamente ");
    if (!confirmar) return;

    try {
        await api(`/clientes/${id}/inativar`, {
            method: "put"
        });

        mostrarMensagem("Cliente inativado com sucesso!", "sucesso");

        buscar();

    } catch (error) {
        mostrarMensagem("Erro ao inativar cliente", "erro");
        console.error(error);
    }
}

File: tests/helpers/clientes.py
from helpers.auth import headers

def criar_cliente(client, token):

    response = client.post(
        "/clientes",
        headers=headers(token),
        json={
            "nome": "Cliente Teste",
            "email": "cliente@email.com",
            "telefone": "11999999999"
        }
    )

    return response.json()["dados"]

def criar_cliente_com_nome_email_telefone(client, token, nome, email, telefone):

    response = client.post(
        "/clientes",
        headers=headers(token),
        json={
            "nome": nome,
            "email": email,
            "telefone": telefone
        }
    )

    return response

def buscar_cliente_por_id(client, token, cliente_id):

    response = client.get(
        f"/clientes/{cliente_id}",
        headers=headers(token)
    )

    return response.json()["dados"]

def buscar_cliente_por_nome(client, token, nome):

    response = client.get(
        f"/clientes?nome={nome}",
        headers=headers(token)
    )

    return response.json()["dados"]

def listar_clientes_ativos(client, token):

    response = client.get(
        "/clientes",
        headers=headers(token)
    )

    return response.json()["dados"]

def atualizar_cliente(client, token, nome, email, telefone):

    cliente = criar_cliente(client, token)

    response = client.put(
        f"/clientes/{cliente['id']}",
        headers=headers(token),
        json={
            "nome": nome,
            "email": email,
            "telefone": telefone
        }
    )

    return response

def inativar_cliente(client, token, cliente_id):

    response = client.put(
    f"/clientes/{cliente_id}/inativar",
    headers=headers(token)
    )

    return response.json()["dados"]

def reativar_cliente(client, token, cliente_id):

    response = client.put(
        f"/clientes/{cliente_id}/reativar",
        headers=headers(token)
    )

    return response.json()["dados"]

def listar_clientes_inativos(client, token):

    response = client.get(
        "/clientes/inativos",
        headers=headers(token)
    )

    return response.json()["dados"]